### PR TITLE
Add section about browser tabs sleeping

### PR DESCRIPTION
**Pull request by BrennanConroy**
[Internal review link](https://review.docs.microsoft.com/en-us/aspnet/core/signalr/javascript-client?view=aspnetcore-6.0&branch=pr-en-us-22773#browser-sleeping-tab)

### CODE REVIEW

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:304]: Best to localize code comments.

```suggestion

For the preceding code example:

* Web Locks are experimental. The conditional check confirms that the browser supports Web Locks.
* The promise resolver (`lockResolver`) is stored so that the lock can be released when it's acceptable for the tab to sleep.
* When closing the connection, the lock is released by calling `lockResolver`. When the lock is released, the tab is allowed to sleep.
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:289]:
```suggestion
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:292]:
```suggestion
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:301]:
```suggestion
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:271]:
```suggestion
* Playing audio
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:272]:
```suggestion
* Holding a Web Lock
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:273]:
```suggestion
* Holding an IndexedDB lock
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:274]:
```suggestion
* Being connected to a USB device
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:275]:
```suggestion
* Capturing video or audio
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:276]:
```suggestion
* Being mirrored
```

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [inline aspnetcore/signalr/javascript-client.md:277]:
```suggestion
* Capturing a window or display
```

Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:277]:
```suggestion
* Playing audio
* Holding a web lock
* Holding an `IndexedDB` lock
* Being connected to a USB device
* Capturing video or audio
* Being mirrored
* Capturing a window or display
```

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:282]:
```suggestion
If your app wants to avoid being put to sleep then it needs to trigger one of the heuristics that the browser looks at.
```

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:284]:
```suggestion
The following code example shows how to make use of [Web Locks](https://developer.mozilla.org/docs/Web/API/Web_Locks_API) to keep the tab awake and avoid unexpected connection closures from the tab going to sleep.
```

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:289]:
```suggestion
// Web locks are experimental,  so verify the browser supports them.
```

**Review by guardrex — APPROVED** [review]
Just nits ... edit suggestions directly if that's easier to merge updates.

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:292]:
```suggestion
        // Store promise resolver so the lock can be released whenever it's OK for the tab to sleep.
```

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:301]:
```suggestion
// When closing the connection release the lock to allow the tab to sleep.
```

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:266]: I have two suggestions for better MT (machine translation). Better or production

```suggestion
A better implementation would use an exponential back-off or retry a specified number of times before giving up.
```

Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:266]:
```suggestion
A production implementation would use an exponential back-off or retry a specified number of times before giving up.
```

**Review by bradygaster — APPROVED** [review]

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [reply aspnetcore/signalr/javascript-client.md:289]: @guardrex suggestion to delete comments in code and add them to the MD file is probably better, but either approach is OK.

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:300]:
```suggestion
```

This is only good if the code comment is inline, removing and relying on the comment below for this.

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:306]:
```suggestion

## Additional resources
```

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:286]:
```suggestion
```

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:268]:
```suggestion
## Browser sleeping tab
```

This was under the "Reconnect clients" section, not sure if it belongs under its own heading, but reconnect doesn't seem correct

**Review by bradygaster — APPROVED** [review]

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:268]: Can we add a non-visible comment here that says there is code in SignalR referencing this anchor so we know not to move it?

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:267]:
```suggestion
<a name="bsleep"></a>

```

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [reply aspnetcore/signalr/javascript-client.md:268]: That's why I like permanent links. Can you accept my suggest

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [reply aspnetcore/signalr/javascript-client.md:267]: Do the docs also support comments `// this is used by code in SignalR, don't rename or remove`?

Comment (guardrex) [reply aspnetcore/signalr/javascript-client.md:267]: As HTML ... `<!-- blah blah blah -->`

Comment (guardrex) [reply aspnetcore/signalr/javascript-client.md:267]: ... the bookmark thing is under ... as Ryan would say ... "civil debate" 😆 lol ... `name` is depreciated under HTML5, and that's our doctype.

Comment (Rick-Anderson) [reply aspnetcore/signalr/javascript-client.md:267]: Like real code, people skip comments, that's why I like the shorter permanent link. We can add both.

Comment (BrennanConroy) [reply aspnetcore/signalr/javascript-client.md:267]: Ah of course, html comments 👍 
```suggestion
<!-- This anchor is used by code in the SignalR Typescript client, do not rename or remove without considering the impacts there -->
<a name="bsleep"></a>

```

Comment (guardrex) [reply aspnetcore/signalr/javascript-client.md:267]: The following is valid ...

```html
<h2 id="bsleep">Browser sleeping tab</h2>
```

... but markdown linting isn't good for it. ....... and docs management might send a hit squad to your house! 😨 (but I had to do it in a couple of pivoting docs to avoid broken bookmarks)

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:272]:
```suggestion
```

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:267]:
```suggestion

<!-- This heading is used by code in the SignalR Typescript client, do not rename or remove without considering the impacts there -->
```

File: aspnetcore/signalr/javascript-client.md
Comment (BrennanConroy) [inline aspnetcore/signalr/javascript-client.md:270]: @richlander @guardrex Can you give this a quick final once over and then I'll merge

File: aspnetcore/signalr/javascript-client.md
Comment (Rick-Anderson) [inline aspnetcore/signalr/javascript-client.md:266]:
```suggestion
Production implementations typically use an exponential back-off or retry a specified number of times.
```

**Review by Rick-Anderson — APPROVED** [review]
One suggestion, looks great and much needed.

File: aspnetcore/signalr/javascript-client.md
Comment (guardrex) [reply aspnetcore/signalr/javascript-client.md:270]: I defer to @Rick-Anderson on this. I only use this when I must; and I think with *version by file* now :tada:, it will only be needed in Blazor docs that use a WASM/Server pivot with duplicate headings.

**Comment by guardrex** [comment]
@BrennanConroy ... Your JS ... two-spaces tho for the indents is our standard AFAIK.

**Comment by guardrex** [comment]
**_Toooo late!_** Oh, well. I think actually that it probably isn't completely standardized around the doc set.

**Pull request merged**
